cabal-install/src/Distribution/Client/IndexUtils.hs

```diff
@@ -480,12 +462,32 @@ readRepoIndex verbosity repoCtxt repo idxState =
           return (mempty, mempty, emptyStateInfo)
         else ioError e
 
+    isOldThreshold :: Double
     isOldThreshold = 15 -- days
-    warnIfIndexIsOld dt = do
-      when (dt >= isOldThreshold) $ case repo of
-        RepoRemote{..} -> warn verbosity $ errOutdatedPackageList repoRemote dt
-        RepoSecure{..} -> warn verbosity $ errOutdatedPackageList repoRemote dt
-        RepoLocalNoIndex{} -> return ()
+    warnIfIndexIsOld isi = do
+      curTime <- getCurrentTime
+      case timestampToUTCTime (isiHeadTime isi) of
+        Nothing -> pure ()
+        Just ts -> do
+          let timeDiffInDays =
+                realToFrac (curTime `diffUTCTime` ts) / realToFrac posixDayLength
+          when (timeDiffInDays >= isOldThreshold) $ case repo of
+            RepoRemote{..} ->
+              warn verbosity $ errOutdatedPackageList repoRemote timeDiffInDays
+            RepoSecure{..} ->
+              warn verbosity $ errOutdatedPackageList repoRemote timeDiffInDays
+            RepoLocalNoIndex{} -> return ()
+
+    dieIfRequestedIdxIsNewer isi =
+      let latestTime = isiHeadTime isi
+       in case idxState of
+            IndexStateTime t -> when (t > latestTime) $ case repo of
+              RepoRemote{..} ->
+                die' verbosity $ errRequestedIdxIsNewer repoRemote latestTime t
```

Comment:
In `getSourcePackagesAtIndexState` we warn the user that index-state is ignored for old-format repositories, so I believe this should be a `pure ()`. I wish types would help :-/

Reply:
Done